torch/csrc/jit/codegen/cuda/root_domain_map.cpp

```diff
@@ -1075,6 +1097,85 @@ bool ComputeAtRootDomainMapBuilder::safeToMap(const DomainKeySet& domains) {
   return true;
 }
 
+namespace {
+class ExactRootDomainMapBuilder : private IterVisitor {
+ public:
+  ExactRootDomainMapBuilder(
+      Fusion* fusion,
+      DisjointSets<const IterDomain*>& eq_sets)
+      : eq_sets_(eq_sets) {
+    traverseFrom(fusion, fusion->outputs());
+  }
+
+ private:
+  using IterVisitor::handle;
+
+  void handle(Expr* expr) final {
+    for (auto producer : ir_utils::filterByType<TensorView>(expr->inputs())) {
+      for (auto consumer :
+           ir_utils::filterByType<TensorView>(expr->outputs())) {
+        PairwiseRootDomainMap pwise_map(producer, consumer, true);
+        const auto mappings = pwise_map.mapProducerToConsumer(
+            producer->domain(), consumer->domain());
+        for (const auto& mapping : mappings) {
+          eq_sets_.mapEntries(mapping.first, mapping.second);
+        }
+      }
+    }
+  }
+
+ private:
+  DisjointSets<const IterDomain*>& eq_sets_;
+};
+
+} // namespace
+
+ExactRootDomainMap::ExactRootDomainMap(Fusion* fusion) {
+  ExactRootDomainMapBuilder builder(fusion, eq_sets_);
+}
+
+bool ExactRootDomainMap::areMapped(
+    const IterDomain* id_a,
+    const IterDomain* id_b) const {
+  return eq_sets_.permissiveAreMapped(id_a, id_b);
```

Comment:
Is permissive required? I would think it doesn't matter. Just seems odd to use `permissive` in `Exact`

Reply:
The difference between `strict` and `permissive` in `DisjointSets` is whether unknown IterDomains are considered an error or not. Querying an exact map with an IterDomain that didn't exist when creating the map, for example, would cause an error with the strict map. I didn't think it's necessary to make it an error as it just queries if it's mapped or not, so thought the permissive version should be fine.

That said, this means that `!ExactRootDomainMap::areMapped(id0, id1)` does not necessarily mean `id0` and `id1` are guaranteed not to be mapped, so it seems it's safer to use the strict version. I'll update the code.